src/cargo/core/compiler/standard_lib.rs

```diff
@@ -158,6 +161,47 @@ pub fn resolve_std<'gctx>(
         HasDevUnits::No,
         crate::core::resolver::features::ForceAllTargets::No,
     )?;
+
+    // Verify that we have resolved to a subset of the lockfile
+    let lockfile = std::fs::read_to_string(&src_path.join("Cargo.lock"))
+        .expect("Couldn't read the Rust source's lockfile");
+
+    let encoded_lockfile: EncodableResolve = toml::from_str(&lockfile).unwrap();
+    let lockfile_packages = encoded_lockfile.package().expect("libstd has no packages!");
+
+    for resolved_pkg in resolve.targeted_resolve.iter() {
+        let pkg_name = resolved_pkg.name().to_string();
+        let pkg_ver = resolved_pkg.version().to_string();
+        let lockfile_pkg = lockfile_packages.binary_search_by_key(
+                &(pkg_name.as_str(), pkg_ver.as_str()),
+                |p| (p.name(), p.version())
+            )
+            .and_then(|idx| Ok(&lockfile_packages[idx]))
+            .unwrap_or_else(|_|
+                panic!("Standard library's package graph changed during resolution:
+                       Package '{}' ({}) was present in the resolve but not in the original lockfile",
+                       resolved_pkg.name(), resolved_pkg.version())
+            );
+        // If the lockfile wasn't sorted then the binary search result can be nonsensical
+        assert!(lockfile_pkg.name() == pkg_name);
+
+        for (dep, _) in resolve.targeted_resolve.deps(resolved_pkg) {
+            lockfile_pkg.dependencies()
+                .and_then(|deps| {
+                    deps.iter().find(|pkg| {
+                        pkg.name() == dep.name().as_str()
+                        && (pkg.version() == None
+                            || pkg.version() == Some(dep.version().to_string().as_str()))
+                    })
+                })
+                .unwrap_or_else(||
+                    panic!("Standard library's package graph changed during resolution:
+                           Package '{}' ({}) was present as a dependency of '{} ({}) in the resolve but not in the lockfile",
+                           dep.name(), dep.version(), resolved_pkg.name(), resolved_pkg.version())
+                );
```

Comment:
Do we need to check the dependencies?  I would assume that so long as the package is deemed the same, the dependencies should be the same.

@Eh2406 in case you know of a reason otherwise.

Reply:
Say rustc and std depend on the same package, but rustc enabled an optional dependency (a). I believe they'd resolve to the same package (with an edge to a) in the lockfile, but the std resolve in build-std would be free to eliminate the edge to a.

This is fine in the patch since the edge exists in the lockfile but not in the resolve - the reverse situation is an error.

Reply:
My first thought was that maybe we should have a comment that we are checking for extra features being activated. As an alternative, my second thought is to update the panic to clarify that.